src/libostree/ostree-repo-pull.c

```diff
@@ -856,24 +856,55 @@ fetch_ref_contents (OtPullData                 *pull_data,
                     GCancellable               *cancellable,
                     GError                    **error)
 {
-  g_autofree char *filename = NULL;
+  g_autofree char *ret_contents = NULL;
 
-  if (ref->collection_id == NULL || g_strcmp0 (ref->collection_id, main_collection_id) == 0)
-    filename = g_build_filename ("refs", "heads", ref->ref_name, NULL);
+  if (pull_data->remote_repo_local != NULL && ref->collection_id != NULL)
+    {
+#ifdef OSTREE_ENABLE_EXPERIMENTAL_API
+      g_autoptr(GHashTable) refs = NULL;  /* (element-type OstreeCollectionRef utf8) */
+      if (!ostree_repo_list_collection_refs (pull_data->remote_repo_local,
+                                             ref->collection_id, &refs,
+                                             OSTREE_REPO_LIST_REFS_EXT_NONE,
+                                             cancellable, error))
+        return FALSE;
+
+      ret_contents = g_strdup (g_hash_table_lookup (refs, ref));
```

Comment:
(See below, this reminds me we need a collection equivalent of `ostree_repo_resolve_rev()`.)

Reply:
I’ll add a basic version of that in this PR, but it’ll basically be a splitting out of the code which is already in the PR. It can be optimised in a follow-up issue.

Reply:
Yep, fine by me.